src/amqp.lisp

```diff
@@ -845,3 +847,428 @@ retrieved has been processed"
+  (cffi:with-foreign-slots ((delivery-tag multiplep requeuep)
+                            method-struct
+                            (:struct amqp-basic-nack-t))
+    (make-method-frame-basic-negative-acknowledgment channel
+                                                      delivery-tag
+                                                      (> multiplep 0)
+                                                      (> requeuep 0))))
+
+(defmethod foreign-method->class ((method-id (eql #.+amqp-basic-return-method+))
+                                  method-struct
+                                  channel)
+    "Returns fresh object of METHOD-FRAME-BASIC-RETURN filled with data from foreign METHOD-STRUCT.
+
+Arguments:
+
+METHOD-ID => CONSTANT, AMQP basic return method id.
+METHOD-STRUCT => CFFI FOREIGN POINTER, decoded basic return method frame.
+CHANNEL => INTEGER, Rabbitmq channel."
+  (cffi:with-foreign-slots ((reply-code reply-text exchange routing-key)
+                            method-struct
+                            (:struct amqp-basic-return-t))
+    (make-method-frame-basic-return channel
+                                    reply-code
+                                    (bytes->string reply-text)
+                                    (bytes->string exchange)
+                                    (bytes->string routing-key))))
+
+(defmethod foreign-method->class ((method-id (eql #.+amqp-basic-deliver-method+))
+                                  method-struct
+                                  channel)
+      "Returns fresh object of METHOD-FRAME-BASIC-DELIVER filled with data from foreign METHOD-STRUCT.
+
+Arguments:
+
+METHOD-ID => CONSTANT, AMQP basic deliver method id.
+METHOD-STRUCT => CFFI FOREIGN POINTER, decoded basic deliver method frame.
+CHANNEL => INTEGER, Rabbitmq channel."
+  (cffi:with-foreign-slots ((consumer-tag delivery-tag redeliveredp exchange routing-key)
+                            method-struct
+                            (:struct amqp-basic-deliver-t))
+    (make-method-frame-basic-deliver channel
+                                     (bytes->string consumer-tag)
+                                     delivery-tag
+                                     (> redeliveredp 0)
+                                     (bytes->string exchange)
+                                     (bytes->string routing-key))))
+
+(defmethod foreign-method->class (method-id method-struct channel)
+      "Returns t when METHOD-ID equals CONFIRM-SELECT-OK-METHOD otherwise signals and error.
+
+Arguments:
+
+METHOD-ID => INTEGER, AMQP basic return method id.
+METHOD-STRUCT => CFFI FOREIGN POINTER, decoded method frame.
+CHANNEL => INTEGER, Rabbitmq channel."
+  (declare (ignore method-struct channel))
+  (when (null (find-if #'(lambda (value)
+                           (eql var value))
+                       '(+amqp-confirm-select-ok-method+)))
+    (error "Unhandled method frame, ID ~s" method-id))
```

Comment:
I would suggest declaring a dedicated condition for this.

Reply:
Done